### DISCUSSION TITLE
hybrid Algo trading interactive UI

### DISCUSSION THREAD

**Discussion by MicroTrendsTom** [MicroTrendsLtd/NinjaTrader8#52] [Show and tell]
**Realtime UI controls  to allow Hybrid Algo mode - manual - man and machine.**
1) Hybrid UI controls to control order entry and exits in WPF is a very basic thing 
2)The complex issue is realtime config to allow multi tabbed view of properties of strategy in PropertyGrid using attribute filters without resorting to making concrete classes to view -

 help wanted with that part and ofcourse collaboration on all and input welcome -see this  section for the implementation:
 https://github.com/MicroTrendsLtd/NinjaTrader8/discussions/52#discussioncomment-2225952

**UI Examples**
This is a popular product for NT7 For AFT7 -  AFT8 the nt8 version to come later- WPF databinding is far superior for sure than WinForms
https://algofuturestrader.zendesk.com/hc/en-us/articles/360055568611-NinjaTrader-7-AFT7-day-trading-futures-system-overview-of-components

Scalper high tech prop trading
https://microtrends.co/wp-content/uploads/2016/11/automated-trading-systems-development-1.png
![image](https://user-images.githubusercontent.com/24366913/153766781-7a77ca7f-374b-401d-92ac-26835cc12c6c.png)

basic plain vanilla
https://microtrends.co/wp-content/uploads/2017/01/NinjaTrader7-BWT-Precision-AutoTrader-7.png
![image](https://user-images.githubusercontent.com/24366913/153766772-c68845a8-90b3-46cf-8cd7-5b061e8174ac.png)

NT8 version
![image](https://user-images.githubusercontent.com/24366913/153766797-b242ed5f-a421-4a36-8cb5-301a74357573.png)

Config panels
Winforms
![image](https://user-images.githubusercontent.com/24366913/153766757-1568b537-6a42-418e-9442-74964af72beb.png)

NT8 propertygrid
![image](https://user-images.githubusercontent.com/24366913/153766803-7f5e12f7-b0ff-4be3-903e-bca0a9a16e30.png)

This is the biggest challenge  -to try make a multi tab view without creating physicals classes e.g. filter by properties data annotations attributes would be the best route  - help wanted on that part for sure...

**Comment by MicroTrendsTom** [comment]
The end result is this so far as of 2022/02/18:
**Hybrid Algo Realtime UI**
![image](https://user-images.githubusercontent.com/24366913/154787269-5abd3018-1ea9-414c-b4a2-bf5225ab2c15.png)

**Realtime Settings Config**
Multi Tab would be much better to have a view of parts of the system for easier access and navigation...
![image](https://user-images.githubusercontent.com/24366913/154787305-f8380017-c3e9-4736-bc92-e2d9312f5104.png)

**Comment by MicroTrendsTom** [comment]
**Popup Config to allow realtime editing of properties of a running system**
This is the biggest challenge -to try make a multi tab view without creating physicals classes e.g. filter by properties data annotations attributes would be the best route - help wanted on that part for sure...
![image](https://user-images.githubusercontent.com/24366913/154787520-c3c0b0a8-7322-455f-9255-f7e00350d113.png)

**Reply by jmscraig** [reply]
"This is the biggest challenge -to try make a multi tab view without creating physicals classes e.g. filter by properties data annotations attributes would be the best route - help wanted on that part for sure..."


![Addon-Tabed](https://user-images.githubusercontent.com/15313578/154996178-6e301ba7-8846-47c6-9cfb-e16b0e58dd04.JPG)

I don't really like idea as it feels still too complex in a number of ways..  but if we are brainstorming..   

Possibly (or not) the simplest way to build a tabbed window in NT's WPF framework would be to build a standalone multi-tabbed windows where in effect each window is expected to be its own set of classes but the independence of the windows might help provide simpler development and supportability.

**Comment by jmscraig** [comment]
An alternative idea.   Still just brainstorming random ideas.    Somewhere I have an example from the NT forums of collapsible categories in Indicator property windows allow expansion similar to your NT7 product's category expansion in this window but with the addition of being still a WPF window and providing all the WPF input controls, button controls etc.     So I am wondering it that might be a paradigm that would be a simpler replacement for using tabs?  

![Categories](https://user-images.githubusercontent.com/15313578/154997695-269342f4-5058-4cf3-84dc-658eee1402ed.JPG)

**Reply by MicroTrendsTom** [reply]
Yes the property gird does this already - Parameter Groups... etc
the tabbed view allows easier view - so a filtered view of the property grid is one possible area of exploration

**Comment by jmscraig** [comment]
Personally i have a long term strategic view for extensibility beyond the narrow markets NT wants to support so bias toward starting to build a migration path from as much dependency on just the NT environment.   

We should also bear in mind that at some point NT will migrate from WPF..  to WinUI3?  Maui?   A UI less microsoft dependent?

**On one hand** these windows below are very NT trade execution centric and right now I am hesitant to migrate to something from WPF until NT reveilles their next step UI plans, which could be just a month or as long as three years from now. 
![NT7-Windows-25](https://user-images.githubusercontent.com/15313578/155002163-879a9f40-dc48-4096-bc77-893bf0ccc3c7.JPG)

 
**On the other hand** I might start a migration now from WPF ... given 1) the global transition from WPF and 2) the NT acquisition of Tradovate.

For example possibly NT will pull a MulitCharts move  (Cancel future development of the MultiCharts Microsoft based product and transfer all new business investment into a build cloud based service instead >> TradeView.com .  

NT's newly owned Tradovate UI looks JavaScript - React centric and extensible, scalable as a cloud SaaS product offering.  

https://trader.tradovate.com/try-it?space=EMicro
![Tradovate-25](https://user-images.githubusercontent.com/15313578/155003245-e77bb22a-8c19-4d12-a9fb-7b67dab7e1c6.JPG)
.

========================================================

Leveraging my historic strategic "reading the tea leaves" and architecture planning skills ... 

### In light of the above .. I would bias to:  

**1)  keep all new dependence on WPF designed with quality, but with a strong bias to simplicity so it is easier to migrate from.

2) In order to avoid the significant complexity of using tabbed windows WPF class or highly interdependent classes build 'out-of-high-speed-trade-execution-path' windows like this one as popup flying windows built with vendor platform independent Javascript-React or WinUI3 or Maui.**  If those approaches fail in the short term then my first thought for simplicity in managing WPF short term would be to build configuration or non-trade-execution-path centric windows as popup WPF windows like the Addon example windows a few posts higher in the thread.

![Config-Window-25](https://user-images.githubusercontent.com/15313578/155002302-a5a0715d-70ee-4259-a7d3-9cbc510386e1.JPG)

**Reply by MicroTrendsTom** [reply]
So the focus for this project will be to build out these as shown here in WPF for NT8 and  later NT9 will require adjustments no doubt.
https://github.com/MicroTrendsLtd/NinjaTrader8/discussions/52#discussioncomment-2208805
 
 ATSQuadrobase UI NinjaBuddy
![image](https://user-images.githubusercontent.com/24366913/155076476-15575234-94ab-4363-a383-7f34e8709772.png)
So   the UI provided here will be the basic one provided in the Hybrid layer -  it maybe in time some tabbed view for further controls such as order types and so on.

Config Realtime settings
A separate window as current which allows non modal side by side settings browsing and editing.
![image](https://user-images.githubusercontent.com/24366913/155076554-fc71a24f-ccf7-4827-a500-3aff7203b161.png)
Additional to come load and save json parameters and skip the look and feel params... leave that to NT XML templates etc

**The wanted help with part is:**
To explore the possibility of filtering the property Grid view by parameter group etc... 
eg hiding properties then a tabbed easy to navigate view could be used  or a vertical grid view
the ability  e.g some ide of what that might look like. tabs with property grid filtered by a group of params attributes etc
![image](https://user-images.githubusercontent.com/24366913/155076428-6c4a6b1a-e466-489a-93fe-c7c301ea7c39.png)
So the way that was implemented was via classes and not filtering.... really not a good option for generics. 

NT9
As for NT roadmap i believe from my last chat with product management .net core WPF is the way ahead for NT9
Certainly WPF has been continued as the next 10 year plan from the 2018 dateline.

but with the advent of Maui and their  Tradovate web business ofcourse we can expect something good to come there too

**Reply by jmscraig** [reply]
Ok.  Sounds good.  Rolling off to sleep.

**Comment by jmscraig** [comment]
Hi Tom,   I am still poking around on WPF tab simplification options. 

Most people view the separate class per tab as a design and test simplification principle.  I have not yet surfaced a superclean approach to build out that Tabbed UI in one class. 

Sharing one option thus far..    This is "Tab Centric" development  option or kit I have seen to simplify tabs.   Not exactly what you were asking for but potentially an option.  

Open Source with the highly usable MIT license.  

https://dragablz.net/
https://github.com/ButchersBoy/Dragablz
His main view says it was updated as recently as 20 days ago.

James

The modern look of this example is what drove me to share this with you as an option.
![Tabs](https://user-images.githubusercontent.com/15313578/156942798-20c222d8-cd86-4efd-83a8-16e82418a9c2.jpg)

**Reply by jmscraig** [reply]
Possibly the best way to simplify building a stand alone WPF window is to use a tops-down kit that might leave complex code but in theory simplify the development and support processes.   

Some times adoption of these kits brings a lot more overhead and constraints than they remove but I decided to share the link and let you be the judge. 

This image from Gong's kit caught my eye.  

The OSS License is a bit more constraining but might work.   https://github.com/punker76/gong-wpf-dragdrop/blob/develop/LICENSE

https://github.com/punker76/gong-wpf-dragdrop

![Gong Tabs](https://user-images.githubusercontent.com/15313578/156943124-f5cf2cbd-6d67-46b5-a740-8066b9e8b1b4.png)

**Comment by MicroTrendsTom** [comment]
ok gotcha interesting   - im definetly keen to use intrinsic WPF controls and  the NTPropertyGrid control as the only 3rd party control we are stuck with. 

Of course people can add their own layers and UI elements for the source here the GUI design is definitely ready to our concrete for now with the following intrinsic WPF controls and the NTPropertyGrid controls:

1) NinjaBuddy 
![image](https://user-images.githubusercontent.com/24366913/156973120-50a192fe-4a53-464d-8e15-dbe85c657145.png)

2) popup config view:
![image](https://user-images.githubusercontent.com/24366913/156973163-1ee4dd64-b38b-4d60-ae05-469a33b4f292.png)

Assistance required:
The only assistance needed so far on this project GUI elements is "NinjaTrader Property Grid" , how to filter its view by a group of property attributes within its SelectedObject.     
There is no issues with adding tabs or elements in tabs - it is the "property grid" behaviors or lack of which  poses the challenge etc - and no open source either etc.
 
For the sample below i had to create data classes for each view shown in the tab and thus the property grid selected object was the data class instance 

 ![image](https://user-images.githubusercontent.com/24366913/156973075-fefda313-0d72-415a-9053-0147f44f44e0.png)


using a filtered view will present an easier UI to the end user for some cases.
Plus this way we can  stick to a generic scenario and not have to add classes and so on etc 

We can simply add a tabbed view of property grid with the specific view as shown above  - 
all of this is very easy to build in WPF and there is no mystery other than the filtered view etc.



might have to make a new discussion for the help assistance wanted on that  if this is to be a generic ongoing discussion on UI
so the assistance part doesnt get lost in the ether etc

**Reply by MicroTrendsTom** [reply]
#64  is now in fact the thread for the assistance on the propertygrid, so we can use this for a general discussion on the UI elements in use etc
